### PR TITLE
[SYCL-MLIR] Add conversion patterns for nd_range, nd_item, item and group operations

### PR DESCRIPTION
**Pull request by victor-eds**
Lower all of the operations in the dialect involving these types.

### CODE REVIEW

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVMPass.cpp [hunk 1]
```diff
@@ -69,6 +70,7 @@ void ConvertSYCLToLLVMPass::runOnOperation() {
 
   ConversionTarget target(*context);
   target.addIllegalDialect<sycl::SYCLDialect>();
+  target.addLegalOp<sycl::SYCLLocalIDOp>();
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVMPass.cpp:73]: Please add a TODO comment to indicate that this is temporary.

**Review by whitneywhtsang — APPROVED** [review]

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVMPass.cpp (hunk 1 above)
Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVMPass.cpp:74]: Done!

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 2]
```diff
@@ -354,6 +419,130 @@ class GetRangeSizePattern : public ConvertOpToLLVMPattern<Op> {
...
+               ConversionPatternRewriter &rewriter) const final {
+    const auto thisArg = adaptor.getOperands()[0];
+    const auto ptrTy = LLVM::LLVMPointerType::get(
+        op.getType(), thisArg.getType()
+                          .template cast<LLVM::LLVMPointerType>()
+                          .getAddressSpace());
+    const auto loc = op.getLoc();
+    const auto dimension = getDimensions(op.getOperand().getType());
+    const auto newValue = [this, loc, thisArg, ptrTy, dimension,
+                           &rewriter]() -> Value {
```
Comment (etiotto) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:447]:
```suggestion
    const Value newValue = [&]() {
```

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 3]
```diff
@@ -354,6 +419,130 @@ class GetRangeSizePattern : public ConvertOpToLLVMPattern<Op> {
...
+        const auto id1 = getID(rewriter, loc, ptrTy, thisArg, 1);
+        const Value prod2 = rewriter.create<arith::MulIOp>(loc, id1, r2);
+        const Value add = rewriter.create<arith::AddIOp>(loc, prod1, prod2);
+        const auto id2 = getID(rewriter, loc, ptrTy, thisArg, 2);
+        return rewriter.create<arith::AddIOp>(loc, add, id2);
+      }
+      default:
+        llvm_unreachable("Invalid number of dimensions");
+      }
+    }();
```
Comment (etiotto) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:477]: Can you remove the lambda and instead just assign the newValue in the various switch cases ?

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 4]
```diff
@@ -354,6 +419,130 @@ class GetRangeSizePattern : public ConvertOpToLLVMPattern<Op> {
...
+    rewriter.replaceOp(op, newValue);
+  }
+};
+
+/// Base pattern for operations building a struct from a SYCL grid operation
+/// \tparam GridOp.
```
Comment (etiotto) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:483]: \tparam -> \t param

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 2 above)
Comment (etiotto) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:447]: Actually can you remove the lambda ?

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 5]
```diff
@@ -354,6 +419,130 @@ class GetRangeSizePattern : public ConvertOpToLLVMPattern<Op> {
...
+      const Value val = rewriter.create<arith::IndexCastUIOp>(
+          loc, structMemberTy, rewriter.create<GridOp>(loc, dimTy, dim));
+      rewriter.create<LLVM::StoreOp>(loc, val, ref);
+    }
+    rewriter.replaceOpWithNewOp<LLVM::LoadOp>(op, alloca);
+  }
+};
+
+/// Base pattern for operations returning the result of a SYCL grid operation
+/// \tparam GridOp.
```
Comment (etiotto) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:526]: \tparam -> \t param

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 6]
```diff
@@ -354,6 +419,130 @@ class GetRangeSizePattern : public ConvertOpToLLVMPattern<Op> {
...
+template <typename Op, typename GridOp>
+class GridOpInitDimPattern : public ConvertOpToLLVMPattern<Op> {
+protected:
+  using ConvertOpToLLVMPattern<Op>::ConvertOpToLLVMPattern;
+
+private:
+  using typename ConvertOpToLLVMPattern<Op>::OpAdaptor;
+
+public:
+  LogicalResult match(Op) const override { return success(); }
```
Comment (etiotto) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:536]: override -> final ?

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 7]
```diff
@@ -354,6 +419,130 @@ class GetRangeSizePattern : public ConvertOpToLLVMPattern<Op> {
...
+  using typename ConvertOpToLLVMPattern<Op>::OpAdaptor;
+  using ConvertOpToLLVMPattern<Op>::ConvertOpToLLVMPattern;
+
+  virtual Value getID(OpBuilder &builder, Location loc, Type ptrTy,
+                      Value thisArg, int32_t index) const = 0;
+  virtual Value getRange(OpBuilder &builder, Location loc, Type ptrTy,
+                         Value thisArg, int32_t index) const = 0;
+
+public:
+  LogicalResult match(Op) const override { return success(); }
```
Comment (etiotto) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:435]: override -> final ?

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 8]
```diff
@@ -354,6 +419,130 @@ class GetRangeSizePattern : public ConvertOpToLLVMPattern<Op> {
...
+private:
+  using GetMemberPattern<Tags...>::getRef;
+  using typename ConvertOpToLLVMPattern<Op>::OpAdaptor;
+  using ConvertOpToLLVMPattern<Op>::getTypeConverter;
+
+protected:
+  using ConvertOpToLLVMPattern<Op>::ConvertOpToLLVMPattern;
+
+public:
+  LogicalResult match(Op) const override { return success(); }
```
Comment (etiotto) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:496]: override -> final ?

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 4 above)
Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:483]: I'm using [`\tparam`](https://www.doxygen.nl/manual/commands.html#cmdtparam) here

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 5 above)
Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:526]: Same as above

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 7 above)
Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:435]: `match` operations are `override` intentionally, as the derived classes might want to override it (see `ItemNoOffsetGetLinearIDPattern`)

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 8 above)
Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:496]: Same as above

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 6 above)
Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:536]: Same as above

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 2 above)
Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp]: Done!

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 7 above)
Comment (etiotto) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:435]: Sure. I know that if a class is expected to be derived the virtual member functions that are expected to be overriden cannot be final.

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 8 above)
Comment (etiotto) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:497]: ok

**Review by etiotto — APPROVED** [review]

**Pull request merged**